### PR TITLE
Use minimum-norm subgradient for clamp/clamp_min/clamp_max at boundary

### LINKED ISSUE #184572: `torch.maximum` / `torch.clamp_min` gradients at tie point violate PyTorch's documented subgradient convention and are inconsistent with `F.relu`
### 🐛 Describe the bug

`torch.maximum(x, tensor(0))`, `torch.clamp_min(x, 0)`, and `torch.clamp(x, min=0)` all implement `max(x, 0)`, but their gradients at `x=0` contradict PyTorch's own autograd documentation and are mutually inconsistent.

## Minimal reproducer

```python
import torch
import torch.nn.functional as F

formulations = [
    ("F.relu(x)",                   lambda x: F.relu(x)),
    ("torch.maximum(x, tensor(0))", lambda x: torch.maximum(x, torch.tensor(0.0))),
    ("torch.clamp_min(x, 0)",       lambda x: torch.clamp_min(x, 0.0)),
    ("torch.clamp(x, min=0)",       lambda x: torch.clamp(x, min=0.0)),
]

for name, fn in formulations:
    x = torch.tensor(0.0, requires_grad=True)
    fn(x).backward()
    print(f"{name:<35} grad at x=0: {x.grad.item()}")
```

## Observed output

```
F.relu(x)                           grad at x=0: 0.0
torch.maximum(x, tensor(0))         grad at x=0: 0.5
torch.clamp_min(x, 0)               grad at x=0: 1.0
torch.clamp(x, min=0)               grad at x=0: 1.0
```

## Expected output

```
F.relu(x)                           grad at x=0: 0.0
torch.maximum(x, tensor(0))         grad at x=0: 0.0
torch.clamp_min(x, 0)               grad at x=0: 0.0
torch.clamp(x, min=0)               grad at x=0: 0.0
```

## Why 0.0 is the correct answer

All four expressions compute the same mathematical function `max(x, 0)`, yet they produce three different gradients at `x=0`. This is a problem in practice: users who refactor between these equivalent formulations silently get different gradient behaviour at zero-initialised weights or ReLU activations that land exactly on zero.

PyTorch's own [autograd documentation](https://pytorch.org/docs/stable/notes/autograd.html) specifies:

> *"If the function is convex (at least locally), use the sub-gradient of minimum norm."*

`max(x, 0)` is convex. Its subdifferential at `x=0` is the interval `[0, 1]`. The element
of minimum norm is **0.0**.

| Formulation | grad at x=0 | Consistent with docs? |
|---|---|---|
| `F.relu(x)` | 0.0 | ✓ |
| `torch.abs(x)` | 0.0 | ✓ |
| `torch.maximum(x, tensor(0))` | 0.5 | ✗ |
| `torch.clamp_min(x, 0)` | 1.0 | ✗ |
| `torch.clamp(x, min=0)` | 1.0 | ✗ |

`F.relu` and `torch.abs` already follow the documented minimum-norm convention. `torch.maximum` and `torch.clamp` do not.

### Versions

```
PyTorch version: 2.13.0.dev20260512+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-37-generic-x86_64-with-glibc2.39
Is CUDA available: True
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 5090
Nvidia driver version: 590.48.01

[pip3] numpy==2.4.4
[pip3] torch==2.13.0.dev20260512+cu130
[pip3] triton==3.7.0+git88b227e2
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by rajfirke**
Fixes #184572

`torch.clamp_min(x, 0)` and `torch.clamp(x, min=0)` return a gradient of 1.0 at `x=0`, violating PyTorch's documented minimum-norm subgradient convention and inconsistent with `F.relu` which returns 0.0:

```python
x = torch.tensor(0.0, requires_grad=True)
torch.clamp_min(x, 0.0).backward()
print(x.grad)  # 1.0 (should be 0.0)
```

As @albanD noted in the issue, the fix is to use strict comparisons (`>` / `<`) instead of non-strict (`>=` / `<=`) in the backward formulas.

**Changes across three layers:**

1. **`tools/autograd/derivatives.yaml`**: `self >= min` to `self > min` for `clamp_min` (Scalar and Tensor variants), `self <= max` to `self < max` for `clamp_max` (Scalar and Tensor variants)
2. **`torch/csrc/autograd/FunctionsManual.cpp`**: Same change in `clamp_backward` for both overloads (Scalar and Tensor min/max)
3. **`torch/_refs/__init__.py`**: `torch.ge` to `torch.gt` and `torch.le` to `torch.lt` in the `clamp` decomposition (torch.compile path)
4. **`test/test_autograd.py`**: `TestClampSubgradient` class verifying gradient is 0.0 at boundary and 1.0 away from boundary for all clamp variants

cc @albanD

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -430,22 +430,22 @@
...
-  self: where(self >= min, grad, at::scalar_tensor(0., grad.options()))
+  self: where(self > min, grad, at::scalar_tensor(0., grad.options()))
   result: auto_element_wise
 
 - name: clamp_min.Tensor(Tensor self, Tensor min) -> Tensor
-  self: where(self >= min, grad, at::scalar_tensor(0., grad.options()))
+  self: where(self > min, grad, at::scalar_tensor(0., grad.options()))
   min: where(self < min, grad, at::scalar_tensor(0., grad.options()))
-  result: where(self_p >= min_p, self_t, min_t)
+  result: where(self_p > min_p, self_t, min_t)
```
Comment (soulitzer) [inline tools/autograd/derivatives.yaml]: When clamp_min's variant accepts Tensor, it is possible for both self and min to accept gradient and so the (0, 0) is no longer a valid subgradient.

We should keep the formulas for the Tensor variants as is and although they are not min-norm always, thats probably fine.

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -2016,16 +2015,16 @@ def clamp(
...
-        # not necessary since `ge` would already `False` when either operands has
+        # not necessary since `gt` would already `False` when either operands has
         # a `nan`. So this line below is redundant
         #   `condition = bitwise_and(condition, bitwise_not(isnan(min)))`
         a = torch.where(condition, a, min)  # type: ignore[arg-type]
     if max is not None:
         a_isnan = torch.isnan(a)
         # same as above, no need to adjust `nan` from `max`
-        condition = torch.bitwise_or(torch.le(a, max), a_isnan)  # type: ignore[arg-type]
+        condition = torch.bitwise_or(torch.lt(a, max), a_isnan)  # type: ignore[arg-type]
```
Comment (soulitzer) [inline torch/_refs/__init__.py]: Why do we need to change the forward logic here?

**Review by soulitzer — CHANGES_REQUESTED** [review]
See comments

**Comment by rajfirke** [comment]
Thank you for the review @soulitzer. Addressed both points:

1. **Tensor variants reverted** — kept `>=`/`<=` for the Tensor overloads since `(0, 0)` is not a valid subgradient when both inputs require grad.
2. **Forward decomposition reverted** — removed the `_refs/__init__.py` change entirely; only the backward formulas (derivatives.yaml + FunctionsManual.cpp) are modified now.

The Scalar variants still use strict `>`/`<` for minimum-norm convention, consistent with relu.

**Comment by rajfirke** [comment]
Hi @soulitzer, the Tensor variant and forward decomposition changes have been reverted per your review. Only Scalar variants now use strict comparisons. Ready for re-review when convenient.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -1238,17 +1238,17 @@ Tensor clamp_backward(
     const Tensor& self,
     const std::optional<Scalar>& min,
     const std::optional<Scalar>& max) {
-  // clamp: gradients not defined on min and max, so we return the subgradient 1
```
Comment (soulitzer) [inline torch/csrc/autograd/FunctionsManual.cpp:1241]: Also revert this one

File: test/test_autograd.py [hunk 4]
```diff
@@ -12154,6 +11814,36 @@ def bernoulli_scalar():
     return torch.tensor(0, dtype=torch.uint8).bernoulli_()
 
 
+class TestClampSubgradient(TestCase):
```
Comment (soulitzer) [inline test/test_autograd.py]: Don't think we need an entirely new test class for this. Can you look at where existing boundary case gradient checks are done, or existing clamp gradient tests?

**Comment by rajfirke** [comment]
Addressed both points @soulitzer:

1. **C++ reverted entirely** — `FunctionsManual.cpp` is untouched now. Only `derivatives.yaml` Scalar variants are changed.
2. **Test inlined into TestAutograd** — removed the standalone class, added `test_clamp_min_max_scalar_subgradient_at_boundary` inside `TestAutograd`.

Final scope: 2 lines changed in `derivatives.yaml` (`>=` to `>` for `clamp_min.Scalar`, `<=` to `<` for `clamp_max.Scalar`) + 1 test method.

**Comment by rajfirke** [comment]
Gentle ping @soulitzer — the Tensor variant and C++ changes were reverted per your second review. Now just 2 lines in derivatives.yaml (Scalar variants only) + 1 test in TestAutograd. Ready for re-review.

**Comment by rajfirke** [comment]
Gentle ping @soulitzer — this is now just 2 lines in derivatives.yaml + 1 test method. Happy to adjust anything further.

**Comment by rajfirke** [comment]
Gentle ping @soulitzer — just 2 lines in derivatives.yaml + 1 test. Ready for re-review whenever convenient.

**Comment by rajfirke** [comment]
Gentle ping @soulitzer — all review feedback addressed: only Scalar variants in derivatives.yaml (2 lines), test inlined into TestAutograd. Approved by @Skylion007. Ready for final review.

**Comment by rajfirke** [comment]
Gentle ping — this has been open for 7 weeks with one approval from @Skylion007. The changes are minimal (2 lines in `derivatives.yaml`, Scalar variants only). Would appreciate a second look from @soulitzer when convenient.

**Comment by albanD** [comment]
Sorry for the churn on this one, it is a pretty subtle change that has deep consequences. 
Taking over in https://github.com/pytorch/pytorch/pull/191142 to do a broader change for all similar functions.

**Comment by rajfirke** [comment]
Thank you @albanD, completely understand, the subgradient convention has broader implications. 

Happy to see this addressed comprehensively in #191142. Closing this PR.
